trunk/man/pkgget.1.in
.TH PKGGET 1 %RELEASE% "" "Linux User's Manual"

.SH NAME

pkgget \- Download files required to build and install packages.

.SH SYNOPSIS

.BI "pkgget [ " options " ] " package_list

.SH DESCRIPTION

.B pkgget
checks that the archives and/or patches required to build and install each
package are present on the system, and downloads any that are not.  Files are
searched for in the source and build directories specified in
.IR "%CONFIG%".
If a file is missing and the package profile lacks a URL for the file, the user
is warned and
.B pkgget
continues with the next file.

Files are downloaded to the current directory.  If this is not the source
directory specified in
.I "%CONFIG%"
then the files should be moved there afterwards.  Download ordering is optimized
so that connections to servers may be used for multiple files.
.BR wget (1)
is used for downloading.

Compression suffixes (e.g.
.IR ".gz" )
are ignored, so if, for example, a
.BR "gzip"ped
file was recompressed with
.B bzip2
after downloading, pkgget will not try to download it again.

.SH PACKAGE LIST

Package profiles may be specified in two ways.  Any combination of the following
options may be used.

.BR -f ", " --profile
.I profile
.BR ...
.RS
Specify files containing package profiles.
.RE

.BR -p ", " --package
.I package
.BR ...
.RS
Specify packages by name.  Profiles will be searched for the profile directory
specified in
.IR "%CONFIG%".
This is the default behaviour if neither option is used.
.RE

.BR --
.RS
Treat remaining arguments as packages or profiles, even if they begin with `-'.
This effectively locks the package list into its current mode.
.RE

.SH OPTIONS

All options must precede
.IR "package_list".

.BR -c ", " --config
.I config_file
.RS
Specify alternative configuration file.
.RE

.BR -l ", " --list
.RS
List download locations for missing files, but do not download them.
.RE

.BR -h ", " -? ", " --help
.RS
Display the help screen.
.RE

.BR -v ", " --version
.RS
Display version information.
.RE

.SH FILES

.I "%CONFDIR%/%CONFIG%"
.RS
The global prometheus configuration file.
.RE

.SH BUGS

If you think you have found a bug or encounter any weird behaviour, please
report it to the address below.

.SH AUTHOR

.RI "-- Gareth Jones <" gareth_jones@users.berlios.de >

.SH SEE ALSO

.BR pkgadd (8),
.BR pkgbuild (1),
.BR pkgcheck (1),
.BR pkginfo (1),
.BR pkginstall (8),
.BR pkgrm (8),
.BR pkgunpack (1),
.BR prometheus.conf (5),
.BR wget (1)


trunk/ChangeLog
2005-06-15 Gareth Jones <gareth_jones@users.berlios.de>

	* x86stepping/Makefile: Added build flags.
	* Makefile.defs: Removed x86stepping build flags.

2005-06-06 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* profiles/lfs-6.1/module-init-tools.pkg: Fixed instructions to ensure
	  that the modprobe.conf man-page does not get rebuilt.

2005-05-31 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* profiles/lfs-6.1/bzip2.pkg: Made use of the version variable.

2005-05-29 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* profiles/lfs-6.1/*.pkg: Removed the (hopefully) last tracks of
	  custom changes to the LFS instructions.

	* profiles/lfs-6.1/linux.pkg: Upgraded to Linux 2.6.11.11.

2005-05-22 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* profiles/lfs-6.1/*: Brought the remaining package profiles up to date
	  wrt to LFS 6.1.

	* profiles/lfs-6.1/linux.pkg: Upgraded to Linux 2.6.11.9.
	* profiles/lfs-6.1/gzip.pkg: Added security fix.
	* profiles/lfs-6.1/glibc.pkg: Added a patch that fixes test suite
	  misbehaviour with Linux 2.6.11.
	* profiles/lfs-6.1/hotplug.pkg: Updated instructions.
	* profiles/lfs-6.1/iproute2.pkg: Replaced the patch with a sed.
	* profiles/lfs-6.1/lfs-bootscripts.pkg: Added.

2005-05-20 Gareth Jones <gareth_jones@users.berlios.de>

	* profiles/gareth: Added base_system.pkg, stage0.pkg, stage1.pkg &
	  stage2.pkg.  Don't want to break lfs-6.1 profiles yet!

	* TODO: Noted fakeroot.
	* Makefile.defs: Consistancy fix (only spacing) & version number
	  updated.

2005-04-02 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* profiles/lfs-6.1/syslog-ng.pkg,
	  profiles/lfs-6.1/sysklogd.pkg: Replaced syslog-ng with sysklogd.
	* profiles/lfs-6.1/linux.pkg: Upgraded to Linux 2.6.11.6, updated
	  patches.
	* profiles/lfs-6.1/findutils.pkg: Upgraded to findutils 4.2.20.
	* profiles/lfs-6.1/udev.pkg: Upgraded to udev 055.
	* profiles/lfs-6.1/util-linux.pkg: Updated patches.
	* profiles/lfs-6.1/bash.pkg: Updated patches.
	* profiles/lfs-6.1/linux-libc-headers.pkg: Upgraded to
	  linux-libc-headers 2.6.11.2.
	* profiles/lfs-6.1/glibc.pkg: Upgraded to glibc 2.3.4.
	* profiles/lfs-6.1/m4.pkg: Upgraded to m4 1.4.3.
	* profiles/lfs-6.1/flex.pkg: Updated patches.
	* profiles/lfs-6.1/coreutils.pkg: Updated patches.
	* profiles/lfs-6.1/iproute2.pkg: Upgraded to iproute2-2.6.11-05030,
	  updated patches.
	* profiles/lfs-6.1/e2fsprogs.pkg: Upgraded to e2fsprogs 1.37.
	* profiles/lfs-6.1/readline.pkg: Updated patches.
	* profiles/lfs-6.1/gettext.pkg: Upgraded to gettext 0.14.3.
	* profiles/lfs-6.1/shadow.pkg: Added patch that fixes lastlog breakage.
	* profiles/lfs-6.1/tar.pkg: Added patch that fixes sparse file breakage.

2005-03-11 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* profiles/lfs-6.1/psmisc.pkg: Upgraded to psmisc 21.6.
	* profiles/lfs-6.1/findutils.pkg: Upgraded to findutils 4.2.19.

2005-03-10 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* profiles/lfs-6.1/{binutils,gcc,glibc}.pkg: Finished toolchain
	  profiles.
	* profiles/lfs-6.1/e2fsprofs.pkg: Fixed check_pkg instructions.

2005-02-27 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* profiles/lfs-6.1/*.pkg: Package and profile updates to the Linux
	  kernel, findutils, udev, libtool, automake, m4, glibc, iana-etc,
	  expect, file, readline, gettext, bzip2 and grub.

2005-02-06 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* profiles/lfs-6.1/*.pkg: Added the remaining new packages since LFS 5.
	  Upgraded glibc instructions.

2005-02-04 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* profiles/lfs-6.1/*.pkg: Added most of the new packages since LFS 5.

2005-01-30 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* profiles/lfs-6.1/*.pkg: Updated another batch of packages to LFS 6.1
	  instructions.

2005-01-15 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* profiles/lfs-6.1/*.pkg: Updated the first batch of profiles to
	  LFS 6.1 (testing) instructions.

2004-12-05 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* scripts/pkgrm.in: Optimized removal of empty directories.
	* scripts/pkgrm.in: Changed backup file extension to "pkg.tar.bz2"
	  instead of "bak.tar.bz2".

2004-11-16 Gareth Jones <gareth_jones@users.berlios.de>

	* testsuite/libprometheus/source_config: Added install_log.conf, logdir
	  & prometheus.conf.

	* testsuite/libprometheus/STATUS: Added.

2004-11-06 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* man/pkginfo.1.in: Written.

2004-10-22 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* scripts/pkginfo.in: Written.

2004-10-20 Gareth Jones <gareth_jones@users.berlios.de>

	* scripts/libprometheus.bash.in: Added quotes around echoed variables in
	  logged_failure, source_file, source_config, source_profile, set_stage,
	  unpack, cat_file & location_of.
	* testsuite/libprometheus/log/tnlycf_stderr.txt: Error message updated.
	* testsuite/libprometheus/source_file: Error message updated in
	  invalid_error & none_error.

	* scripts/compact.in: Added quotes around echoed variables.
	* scripts/pkgadd.in: Added quotes around echoed variables.
	* scripts/pkgbuild.in: Added quotes around echoed variables.
	* scripts/pkgcheck.in: Added quotes around echoed variables.
	* scripts/pkgget.in: Added quotes around echoed variables.
	* scripts/pkginstall.in: Added quotes around echoed variables.
	* scripts/pkunpack.in: Added quotes around echoed variables.

2004-09-26 Oliver Brakmann <oliverbrakmann@users.berlios.de>

	* man/*, scripts/*: Changed the short option equivalent of
	  --version to '-V' instead of '-v'.

2004-09-23 Gareth Jones <gareth_jones@users.berlios.de>

	* testsuite/libprometheus/echo_array/echo_array: Added.

	* testsuite/libprometheus/source_file: Added valid.txt, invalid.txt,
	  valid_status, valid_sourced, invalid_status, invalid_error,
	  none_status & none_error.

2004-09-19 Gareth Jones <gareth_jones@users.berlios.de>

	* testsuite/libprometheus/get_installed_version: Added vtab &
	  get_installed_version.


tags/prometheus-0.1/pkgget.sh
#! /bin/bash

# Copyright (C) 2003-2004 Oliver Brakmann <oliverbrakmann@users.berlios.de> &
# Gareth Jones <gareth_jones@users.berlios.de>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
# Place, Suite 330, Boston, MA  02111-1307  USA


# Usage:
#
# pkgget {[-h|-?|--help] | [{-f|--config} <config file>] [-l|--list] <package>
#								[<package> ...]}
# Check if files needed by <package>s are present on system.  Download any that
# are not.

# Global prometheus functions.
source "%LIBPKG%" || exit 1

# Usage: usage
# Print help text.
function usage {
	cat <<-EOF
		Usage:
		$(basename $0) [<options>] <package> [<package> ...]

		Check if any files required by <package>s are present on the
		system and download any that are not.  <package> can be a
		package name, (\`bash'), or a package profile, (\`bash.pkg').
		
		Options:
		-h|-?|--help			Show this help screen
		-f|--config <config file>	Specify alternative configuration file
		-l|--list			List download locations but do not download

	EOF
}

# Usage: remove_found
# Read list of files, output any that are not present in $srcdir or $builddir.
# Warn on any that are missing & cannot be downloaded.
function remove_found {
	local file location protocol
	while read file; do

		# Skip blank lines.
		[[ -z "$file" ]] && continue

		# Skip files that are present.  Suppress not-found warnings.
		location=$(location_of "$file" 2> /dev/null)
		[[ -n "$location" ]] && continue

		# $file not present.  Can we download it?
		protocol="${file%%://*}"
		[[ "$protocol" == "http" || "$protocol" == "ftp" ]] &&
		echo "$file" && continue

		# $file not present & cannot be downloaded.
		warning "$file not found and cannot be downloaded"
	done
}

# Usage: list_missing <package>
# Print missing files in <package>.
function list_missing {
	if (( $# != 1 )) || [[ -z "$1" ]]; then
		usage_error "list_missing" "<package>" "$*"
	fi

	# Run in a subshell to avoid contamination between packages.
	(
		source_profile "$1"
		list_files | remove_found
	)
}

# Usage: split_components
# Read URLs from stdin & output lines with three `:'-separated fields:
# <protocol>:<hostname>:<path>
# <hostname> will be byte-reversed for easy of sorting.
function split_components {
	local file protocol host path
	while read file; do

		# Get & remove $protocol from $file.
		protocol="${file%%://*}"
		file="${file#$protocol://}"

		# Get $host.  Byte-reverse it.
		host="${file%%/*}"
		host=$(echo "$host" | rev)

		# Get $path.
		path="${file#*/}"

		echo "$protocol:$host:$path"
	done
}

# Usage: merge_components
# Read lines from stdin containing three `:'-separated fields:
# <protocol>:<hostname>:<path>
# Reconstruct & output the original URLs, byte-reversing <hostname>.
function merge_components {
	local line protocol host path
	while read line; do

		# Get & remove $protocol from $line.
		protocol="${line%%:*}"
		line="${line#$protocol:}"

		# Get $host.  Byte-reverse it.
		host="${line%%:*}"
		host=$(echo "$host" | rev)

		# Get $path.
		path="${line#*:}"

		echo "$protocol://$host/$path"
	done
}

# Usage: sort_downloads
# Read a list of downloads & sort them by server.  This allows downloaders to
# use persistant connections.
function sort_downloads {
	split_components |
	sort -t : -k 1,1 -k 2,2 -k 3,3 |
	merge_components
}

# Usage: list_downloads <package> [<package> ...]
# Sort & print download locations of file missing for <package>s.
function list_downloads {
	(( $# == 0 )) &&
	usage_error "list_downloads" "<package> [<package> ...]"

	for package in "$@"; do
		list_missing "$package"
	done | sort_downloads
}

# --------------------------- START EXECUTION HERE --------------------------- #

# Check for options.
while [[ "$1" == -* ]]; do
	case "$1" in
		(-h|-\?|--help)	usage
				exit 0;;

		(-f|--config)	[[ -z "$2" ]] &&
				script_usage_err "$1: no config file specified"

				pkgconfig="$2"
				shift;;

		(-l|--list)	listonly=yes ;;

		(*)		script_usage_err "unrecognised option: $1"
	esac
	shift
done

# The remaining arguments should all be packages.
(( $# == 0 )) && script_usage_err "no packages to get"

source_config

if [[ "$listonly" == yes ]]; then
	list_downloads "$@"
else
	list_downloads "$@" | wget --input-file=- ||
	error "download failed"
fi
